File: Leetcode/0023-MergeKSortedLists/java/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> pq = new PriorityQueue<>((a, b) -> {
            return a.val - b.val;
        });
        ListNode dummy = new ListNode();
        // Add initial k nodes into pq
        for (ListNode node : lists) {
            if (node != null) {
                pq.add(node);
            }
        }
        ListNode node = dummy;
        while (!pq.isEmpty()) {
            node.next = pq.remove();
            node = node.next;
            if (node.next != null) {
                pq.add(node.next);
            }
        }
        return dummy.next;
    }
}

File: Leetcode/0023-MergeKSortedLists/go/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
import "container/heap"

type MinHeap []*ListNode

func (h MinHeap) Len() int { return len(h) }
func (h MinHeap) Less(i, j int) bool { return h[i].Val < h[j].Val }
func (h MinHeap) Swap(i, j int) {h[i], h[j] = h[j], h[i]}

func (h *MinHeap) Push(x any) {
    node := x.(*ListNode)
    (*h) = append(*h, node)
}

func (h *MinHeap) Pop() any {
    val := (*h)[len(*h)-1]
    (*h) = (*h)[:len(*h)-1]
    return val
}

func mergeKLists(lists []*ListNode) *ListNode {
    dummy := &ListNode{}
    // Add initial K elements into min heap
    h := &MinHeap{}
    heap.Init(h)
    for _, node := range lists {
        if node != nil {
            heap.Push(h, node)
        }
    }
    node := dummy
    for h.Len() > 0 {
        node.Next = heap.Pop(h).(*ListNode)
        node = node.Next
        if node.Next != nil {
            heap.Push(h, node.Next)
        }
    }
    return dummy.Next
}